### ISSUE TITLE
create new c++ scene graph detector tool with real-time preview of bounding boxes and scene graph

### ISSUE THREAD

**Issue by ingojaeckel** [ingojaeckel/go-raspberry-pi-timelapse#207]
# Add `cpp-scene-graph-detector`: C++ Scene Graph (Objects + Relations) Pipeline with CI for macOS (Intel, 2018 MBP) and AMD64 Linux (CPU/GPU)

## 🧭 Goal

Create a new C++ component that takes single images or video frames, detects objects, and produces a **scene graph** (nodes = objects, edges = spatial/semantic relations).  
Design and project layout should mirror the style and conventions of [`cpp-object-detection`](../cpp-object-detection).

Just like `cpp-object-detection`, provide a real-time view-finder (`--show-preview`) showing the live feed of the selected web cam.
* overlay bounding boxes of detected objects
* at the bottom of the screen, overlay a textual description of the currently visible scene graph. 
* each time the scene changes significantly (e.g. new object type detected), record a photo to disk including the overlays (following the behaviour currently in place in `cpp-object-detection`)

---

### 1. Language & Style
- Implement in **C++17** (or newer if already used by the repo).
- Follow the **structure and coding style** of `cpp-object-detection`:
  - CMake-based build
  - Consistent folder layout, naming, logging, and CLI patterns
  - Keep third-party dependencies minimal and pinned

### 2. New Module
- Create a **sibling folder** to `cpp-object-detection`:
  ```
  /cpp-scene-graph-detector
  ```
- Self-contained library + CLI binary (like `cpp-object-detection`).

### 3. Platforms
- **macOS (Intel) — 2018 MacBook Pro**:  
  CPU execution required; GPU acceleration via **OpenCL** is best-effort if available.  
- **Linux x86_64 (AMD64)**:  
  - CPU execution required.  
  - GPU acceleration via **OpenCL** if available (targeting AMD GPUs).  
- Must **build successfully** on both platforms; runtime GPU detection is optional.

### 4. Models (Pretrained Only)
- **Object Detection:**  
  Load an ONNX detector (e.g., YOLO family) via **OpenCV DNN**.  
- **Relation / Predicate Head:**  
  Load a lightweight ONNX model predicting `(subject, predicate, object)` triplets.  
  - Support spatial predicates: `left_of`, `right_of`, `in_front_of`, `behind`, `overlaps`, `contains`, `intersects`, `between`.  
  - Extendable with semantic predicates like `on`, `under`, `next_to`.  
- **No training code** — use pre-exported ONNX models from `assets/models/`.

### 5. Backends
- Inference via **OpenCV DNN**:
  - **CPU** backend (default)
  - **OpenCL** backend (optional)
- One build target; backend selected at runtime.

### 6. Outputs
- **Bounding boxes** with labels + confidences.  
- **Scene Graph:**
  - In-memory representation (`Graph` with `Node` and `Edge`).
  - JSON export (`.json`) and optional Graphviz (`.dot`).
  - Optional ASCII summary printed to stdout.

### 7. CLI
Mirror `cpp-object-detection` UX:

```bash
cpp-scene-graph-detector \
  --input path/to/image_or_video \
  --model.detector assets/models/detector.onnx \
  --model.relations assets/models/relations.onnx \
  --labels assets/labels.txt \
  --backend [cpu|opencl|auto] \
  --threshold.obj 0.25 \
  --threshold.rel 0.5 \
  --max-objects 128 \
  --out.json out/scene_graph.json \
  [--out.dot out/scene_graph.dot] \
  [--visualize out/overlay.jpg]
```

For video: process N frames per second, write rolling outputs.

### 8. Library API (Headers)
```
include/scene_graph/
├── Detector.h      // loadDetector(), infer()
├── RelPredictor.h  // loadRelPredictor(), infer()
├── Graph.h         // Node, Edge, SceneGraph, to_json(), to_dot()
└── Runner.h        // Orchestrates pipeline
```

Example structures:
```cpp
struct Node { int id; int class_id; std::string label; BBox bbox; float score; };
struct Edge { int src_id; int dst_id; int predicate_id; std::string label; float score; };
```

### 9. Performance / Constraints
- Target **~1 FPS** on 2018 MBP CPU at 1280×720.
- Memory footprint within 8–16 GB.
- Deterministic outputs given same inputs.

### 10. Testing
- **Unit tests** (gtest):
  - JSON export schema validation
  - Relation mapping and backend selection
- **Smoke tests:**
  - Load small ONNX models
  - Run on one image → verify valid JSON output
- Include small sample image + golden JSON in `samples/`.

### 11. Artifacts & Assets
- `assets/models/` — placeholder models + README with URLs/checksums  
- `assets/labels.txt` — COCO or custom mapping  
- `docs/SCENE_GRAPH_FORMAT.md` — JSON schema + examples  
- `samples/` — 1–2 images and outputs

### 12. CI (GitHub Actions)
- Add workflows under `.github/workflows/`, mirroring repo style.
- **Build matrix:**
  - `ubuntu-22.04` (amd64)
  - `macos-13` (Intel runner)
- Steps:
  - Install dependencies (OpenCV, gtest)
  - Configure & build with CMake
  - Run unit tests and smoke tests
- Optional job to compile with **OpenCL** enabled.

### 13. Documentation
- `README.md` in `/cpp-scene-graph-detector`:
  - What is a scene graph
  - Build instructions (macOS Intel + Ubuntu AMD64)
  - Running examples
  - JSON schema link
  - GPU/OpenCL notes
- `SCENE_GRAPH_FORMAT.md`:
  - JSON keys: `objects`, `relations`, `meta`
  - Object: `{ id, label, class_id, score, bbox }`
  - Relation: `{ subject_id, predicate, object_id, score }`
  - Example: two houses, one tree, one bush, with `between` relation.

### 14. Optional: Temporal Mode
- Add a simple **tracker** to maintain node IDs across frames.
- IoU-based tracker (no external model).
- If `--temporal` flag is set:
  - Maintain rolling graph.
  - Keep occluded nodes with `visibility=false` for N frames.

---

## 🧩 Acceptance Criteria

- [ ] Builds on **macOS-13 (Intel)** and **Ubuntu 22.04 (amd64)** in CI.  
- [ ] End-to-end CPU pipeline works; CLI demo provided.  
- [ ] JSON scene graph includes ≥3 object classes and ≥3 spatial predicates.  
- [ ] OpenCL backend compiles; runtime gracefully falls back to CPU.  
- [ ] Unit + smoke tests pass in CI.  
- [ ] Code style, logging, and CLI consistent with `cpp-object-detection`.

---

## 🧰 Implementation Notes (for Copilot)

- Use **OpenCV ≥ 4.8** with DNN and OpenCL enabled.  
- Detector & relation heads must be pluggable (interfaces + factories).  
- CMake options:
  - `-DSCENE_GRAPH_WITH_OPENCL=ON|OFF`
  - `-DSCENE_GRAPH_BUILD_TESTS=ON|OFF`
- Include lightweight preprocessing utilities (letterbox, NMS, affine transforms).  
- Keep dependencies local; rely on system packages in CI.

---

## 📦 Deliverables

- `/cpp-scene-graph-detector` source, include, CMake, tests, docs, samples  
- GitHub Actions workflows for macOS and Linux  
- `assets/` README with model download info  
- Working CLI that emits bounding boxes + JSON scene graph
